File: tests/rumina_fs_path_extended_test.rs
use rumina::{Value, run_rumina_with_dir};
use std::fs;

#[test]
fn test_rumina_fs_realpath_link_symlink() {
    let temp_dir = std::env::temp_dir().join("rumina_fs_path_ext_test");
    fs::create_dir_all(&temp_dir).expect("create temp dir failed");

    let src = temp_dir.join("src.txt");
    let hard = temp_dir.join("hard.txt");
    let sym = temp_dir.join("sym.txt");

    let src_s = src.to_string_lossy().replace('\\', "\\\\");
    let hard_s = hard.to_string_lossy().replace('\\', "\\\\");
    let sym_s = sym.to_string_lossy().replace('\\', "\\\\");

    let code = format!(
        r#"
include "rumina:fs";
include "rumina:path";

fs.writeText("{src}", "hello");
fs.link("{src}", "{hard}");
fs.symlink("{src}", "{sym}");

var rp = fs.realpath("{src}");
var lk = fs.readLink("{sym}");
var ents = fs.readDir("{dir}", true);
var first_ok = size(ents) > 0 && ents[0].name != "";

var ok =
  fs.exists("{hard}") &&
  fs.exists("{sym}") &&
  path.basename(rp) == "src.txt" &&
  path.basename(lk) == "src.txt" &&
  first_ok;

if (ok) {{ "ok"; }} else {{ "bad"; }}
"#,
        src = src_s,
        hard = hard_s,
        sym = sym_s,
        dir = temp_dir.to_string_lossy().replace('\\', "\\\\")
    );

    let result = run_rumina_with_dir(&code, Some(temp_dir.to_string_lossy().to_string()))
        .expect("execution failed");

    match result {
        Some(Value::String(s)) => assert_eq!(s, "ok"),
        other => panic!("Expected Some(String), got {:?}", other),
    }

    fs::remove_dir_all(&temp_dir).ok();
}

File: CRumina/tests/rumina_fs_path_extended_test.cc
#include <test_framework.h>
#include <run_rumina.h>
#include <value.h>
#include <fstream>
#include <filesystem>

using namespace rumina;
using namespace rumina::test;

void test_rumina_fs_realpath_link_symlink() {
    std::string temp_dir = create_temp_dir("rumina_fs_path_ext_test");
    
    std::string src = temp_dir + "/src.txt";
    std::string hard = temp_dir + "/hard.txt";
    std::string sym = temp_dir + "/sym.txt";
    
    std::ofstream src_file(src);
    src_file << "hello";
    src_file.close();
    
    std::string code = 
        "include \"rumina:fs\";"
        "include \"rumina:path\";"
        "fs.link(\"" + src + "\", \"" + hard + "\");"
        "fs.symlink(\"" + src + "\", \"" + sym + "\");"
        "var rp = fs.realpath(\"" + src + "\");"
        "var lk = fs.readLink(\"" + sym + "\");"
        "var ents = fs.readDir(\"" + temp_dir + "\", true);"
        "var first_ok = size(ents) > 0 && ents[0].name != \"\";"
        "var ok ="
        "  fs.exists(\"" + hard + "\") &&"
        "  fs.exists(\"" + sym + "\") &&"
        "  path.basename(rp) == \"src.txt\" &&"
        "  path.basename(lk) == \"src.txt\" &&"
        "  first_ok;"
        "if (ok) { \"ok\"; } else { \"bad\"; }";
    
    auto result = run_code_with_dir(code, temp_dir);
    assert_ok(result);
    
    auto value = result.value();
    assert_true(value.has_value());
    assert_eq(value.value().toString(), "ok");
    
    remove_temp_dir(temp_dir);
}

int main() {
    TestRunner runner;
    
    runner.add_test("rumina_fs_realpath_link_symlink", test_rumina_fs_realpath_link_symlink);
    
    return runner.run_all();
}
